### ISSUE TITLE
recursion error on conda create -n my-r-env -c r r-essentials

### ISSUE THREAD

**Issue by jorahn** [conda/conda#1738]
following the [blog post](https://www.continuum.io/blog/developer/jupyter-and-conda-r) on setting up r and jupyter with irkernel through conda, i run into several problems and don't get jupyter notebook to run on osx el capitan. some seem to origin with irkernel ([#206](https://github.com/IRkernel/IRkernel/issues/206), [#205](https://github.com/IRkernel/IRkernel/issues/205), [#176](https://github.com/IRkernel/IRkernel/issues/176), [#165](https://github.com/IRkernel/IRkernel/issues/165)) but as conda is also referenced (rzmq packaging/linking) there and i don't see this mentioned here yet, maybe it's a combination.

here's the error, conda throws:
```bash
> conda create -n my-r-env -c r r-essentials
Fetching package metadata: ......
Solving package specifications: ...............................................................
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/Users/me/anaconda/bin/conda", line 5, in <module>
    sys.exit(main())
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/cli/main.py", line 195, in main
    args_func(args, p)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/cli/main.py", line 202, in args_func
    args.func(args, p)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/cli/main_create.py", line 49, in execute
    install.install(args, parser, 'create')
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/cli/install.py", line 333, in install
    update_deps=args.update_deps)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/plan.py", line 411, in install_actions
    update_deps=update_deps):
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 1057, in solve
    installed, minimal_hint=minimal_hint, update_deps=update_deps)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 809, in solve2
    dists = self.get_dists(specs, filtered=True)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 710, in get_dists
    filtered = self.prune_packages(specs)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 506, in prune_packages
    if sum(touch(fn) for fn in self.find_matches(ms, True)) == 0:
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 506, in <genexpr>
    if sum(touch(fn) for fn in self.find_matches(ms, True)) == 0:
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 496, in touch
    touch(f2, force)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 492, in touch
    val = touched[fn] = force or is_valid(fn)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 483, in is_valid
[...]
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 516, in find_matches
    if self.filter_[fn] and ms.match(fn):
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 325, in match
    return any(vs.match(version) for vs in self.vspecs)
  File "/Users/me/anaconda/lib/python3.4/site-packages/conda/resolve.py", line 325, in <genexpr>
    return any(vs.match(version) for vs in self.vspecs)
RuntimeError: maximum recursion depth exceeded 
```

* os x 10.11.1

```
Current conda install:

             platform : osx-64
        conda version : 3.18.3
  conda-build version : 1.18.1
       python version : 3.4.3.final.0
     requests version : 2.8.1
     root environment : /Users/me/anaconda  (writable)
  default environment : /Users/me/anaconda
     envs directories : /Users/me/anaconda/envs
        package cache : /Users/me/anaconda/pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/osx-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/osx-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : None
    is foreign system : False
```

* r                         3.2.2                         0  
* r-essentials              1.1                    r3.2.2_0  
* r-rzmq                    0.7.7                  r3.2.2_1

**Comment by jorahn** [comment]
btw i got none of these problems installing r through homebrew and jupyter through a standard python.org 3.4 env

**Comment by cfljam** [comment]
I broke ability to run Jupyter r-kernels in a previously  working install of r-essentials  on Linux by running updates that seem to have resulted in an upgrade to libzmq 5 I am unable to reverse. 
Im sure Im not the only one drawn to conda  by the Continuum blog post, impressed by the ability to install the Jupyter+ R stack and now somewhat  disappointed that I didnt RTFD and clone my env before applying updates. 

```
> IRkernel::main()
Error in dyn.load(file, DLLpath = DLLpath, ...) :
  unable to load shared object '/opt/conda/lib/R/library/rzmq/libs/rzmq.so':
  libzmq.so.4: cannot open shared object file: No such file or directory 
```

**Comment by mdekstrand** [comment]
I am experiencing this same problem - maximum recursion depth exceeded when installing (some) R packages, namely r-irkernel, on OS X. I do not have this problem on Linux.

Command line:

    conda create -n rs-asn -c r r r-irkernel

Last several lines of result:

```
File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in is_valid
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 483, in <genexpr>
    if not any(is_valid(f2) for f2 in self.find_matches(ms, True)):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 516, in find_matches
    if self.filter_[fn] and ms.match(fn):
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 325, in match
    return any(vs.match(version) for vs in self.vspecs)
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 325, in <genexpr>
    return any(vs.match(version) for vs in self.vspecs)
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 301, in match
    return all(c.match(version) for c in self.constraints)
  File "/Users/michael/anaconda/lib/python2.7/site-packages/conda/resolve.py", line 301, in <genexpr>
    return all(c.match(version) for c in self.constraints)
RuntimeError: maximum recursion depth exceeded
```

I am running Anaconda for Python 2.7 with Conda 3.18.3 py27_0 on OS X Mavericks.

**Comment by mdekstrand** [comment]
I have injected a `print` statement to try to debug the list that is causing the problem (dumps `find_matches` results), and I see the following:

```
[u'gmp-5.1.2-0.tar.bz2', u'gmp-5.1.2-4.tar.bz2', u'gmp-5.1.2-3.tar.bz2', u'gmp-5.1.2-2.tar.bz2', u'gmp-5.1.2-6.tar.bz2', u'gmp-5.1.2-5.tar.bz2', u'gmp-5.1.3-0.tar.bz2']
[u'isl-0.12.2-1.tar.bz2', u'isl-0.12.2-0.tar.bz2']
[u'mpc-1.0.1-0.tar.bz2']
[u'mpfr-3.1.2-0.tar.bz2']
[u'cloog-0.18.0-0.tar.bz2']
```

Repeated many times.

My interpretation of this is that there is a dependency cycle involving `gmp`, `isl`, `mpc`, `mpfr`, and `cloog`, and that Conda does not detect the dependency cycle to either handle it safely or exit with a clean error message.

**Comment by pavoljuhas** [comment]
I am getting the same infinite-recursion problem when package is present both in my channel and in defaults:
```shell
$ conda install -c pavoljuhas/channel/dev boost
...
  File "/home/juhas-local/arch/miniconda/lib/python2.7/site-packages/conda/resolve.py", line 516, in find_matches
    if self.filter_[fn] and ms.match(fn):
  File "/home/juhas-local/arch/miniconda/lib/python2.7/site-packages/conda/resolve.py", line 325, in match
    return any(vs.match(version) for vs in self.vspecs)
  File "/home/juhas-local/arch/miniconda/lib/python2.7/site-packages/conda/resolve.py", line 325, in <genexpr>
    return any(vs.match(version) for vs in self.vspecs)
RuntimeError: maximum recursion depth exceeded
```

My channel `pavoljuhas` provides boost 1.56.0 and the boost version in defaults is 1.57.0.  To my understanding conda should install the defaults package 1.57.0, but instead fails with an infinite recursion.  

Installation works after adding an explicit version
```shell
$ conda install -c pavoljuhas/channel/dev boost=1.57.0
```
of after downgrading to conda 3.18.2
```shell
$ conda install -n root conda=3.18.2
$ conda install -c pavoljuhas/channel/dev boost
```

**Comment by mdekstrand** [comment]
I can confirm that downgrading to 3.18.2 also fixes the problem for me.

**Comment by mdekstrand** [comment]
Looking at the changelog and a diff, it looks to me like #1702 is likely where the problem got introduced.

**Comment by CathDalzell** [comment]
I have the same problem. I ran
`conda create -n my-r-env -c r r-essentials`
and got the same recursion error:
`RuntimeError: maximum recursion depth exceeded`

I am running Ubuntu 15.04

**Comment by electronwill** [comment]
https://github.com/conda/conda/issues/1748 is the same issue, and has a little more info about work towards a fix.

**Comment by mcg1969** [comment]
Yes, this was fixed in https://github.com/conda/conda/pull/1749, which simply needs to be released. Hopefully very soon.

**Comment by mcg1969** [comment]
This has been fixed in 3.18.4.
